### ISSUE TITLE
TFRecordDataset entries shuffle between examples

### ISSUE THREAD

**Issue by yipsang** [tensorflow/tensorflow#23098]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS Mojave 10.14 (18A391)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.11.0
- Python version: 3.6
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A
- Have I written custom code: N/A
- Exact command to reproduce: Check Code to reproduce the issue

**Describe the problem**
When I load several examples into a TFRecordDataset and parse the serialised examples using `parse_single_example`, the entry values of the examples are interchanged randomly. For example, if I save two examples `{"a": [1, 2, 3], "b": [10, 20, 30]}` and `{"a": [3, 4, 5], "b": [30, 40, 50]}` into a `tfrecord` and then load and parse it using `TFRecordDataset` and `parse_single_example`, the loaded examples will become `{"a": [3, 4, 5], "b": [10, 20, 30]}` and `{"a": [1, 2, 3], "b": [30, 40, 50]}`, while it will be correct if I use `Example.ParseFromString` to parse the evaluated string.

**Code to reproduce the issue**
Code to write the tfrecord
```
import tensorflow as tf

test = tf.train.Features(feature={
    'a': tf.train.Feature(float_list=tf.train.FloatList(value=[1,2,3])),
    'b': tf.train.Feature(float_list=tf.train.FloatList(value=[10,20,30]))
})
test_2 = tf.train.Features(feature={
    'a': tf.train.Feature(float_list=tf.train.FloatList(value=[3,4,5])),
    'b': tf.train.Feature(float_list=tf.train.FloatList(value=[30,40,50]))
})

example = tf.train.Example(features=test)
example_2 = tf.train.Example(features=test_2)

with tf.python_io.TFRecordWriter('test.tfrecord') as writer:
    writer.write(example.SerializeToString())
    writer.write(example_2.SerializeToString())
```
Code to load the tfrecord. Example parsed via `ParseFromString` is printed first, and then the one parsed via `parse_single_example`
```
dataset = tf.data.TFRecordDataset('test.tfrecord')
dataset = dataset.repeat()
examples = dataset.make_one_shot_iterator()
with tf.Session() as sess:
    for _ in range(2):
        serialized_example = examples.get_next()
        correct_example = tf.train.Example()
        correct_example.ParseFromString(serialized_example.eval())
        print(correct_example)
        features = {
            "a": tf.FixedLenFeature([3], tf.float32),
            "b": tf.FixedLenFeature([3], tf.float32),
        }
        example = tf.parse_single_example(serialized=serialized_example, features=features)
        for name, tensor in example.items():
            print('{}: {}'.format(name, tensor.eval()))
```

**Other info / logs**
The log I've got from the second snippet
```
features {
  feature {
    key: "a"
    value {
      float_list {
        value: 1.0
        value: 2.0
        value: 3.0
      }
    }
  }
  feature {
    key: "b"
    value {
      float_list {
        value: 10.0
        value: 20.0
        value: 30.0
      }
    }
  }
}

a: [3. 4. 5.]
b: [10. 20. 30.]
features {
  feature {
    key: "a"
    value {
      float_list {
        value: 3.0
        value: 4.0
        value: 5.0
      }
    }
  }
  feature {
    key: "b"
    value {
      float_list {
        value: 30.0
        value: 40.0
        value: 50.0
      }
    }
  }
}

a: [1. 2. 3.]
b: [30. 40. 50.]
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
Exact command to reproduce

**Comment by feihugis** [comment]
@yipsang @harshini-gadige I think this is not caused by a bug in TensorFlow. Instead, there are some issues in the example code. In `serialized_example = examples.get_next()`, `serialized_example` is an iterator, but both `correct_example` and `example` will call `get_next` in one loop. That is why the output is not in the right order. For reading part, the below code can return the right output:
```
    dataset = tf.data.TFRecordDataset(filename)
    examples = dataset.make_one_shot_iterator()
    with tf.Session() as sess:
        for _ in range(2):
            serialized_example = examples.get_next()
            # correct_example = tf.train.Example()
            # correct_example.ParseFromString(serialized_example.eval())
            # print(correct_example)
            features = {
                "a": tf.FixedLenFeature([3], tf.float32),
                "b": tf.FixedLenFeature([3], tf.float32),
            }
            example = tf.parse_single_example(serialized=serialized_example, features=features)

            keys = []
            values = []
            for name, tensor in example.items():
                keys.append(name)
                values.append(tensor)

            real_values = sess.run(values)
            for i in range(len(keys)):
                print('{}: {}'.format(keys[i], real_values[i]))
```

**Comment by harshini-gadige** [comment]
@shivaniag  -  Hi, could you please look into this ?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mrry** [comment]
Thanks @feihugis for diagnosing the problem! The iterator is working as intended, and the problem comes from comparing the values returned from different `Session.run()` calls.
